### ISSUE TITLE
Phone numbers in scheduled messages only update in real time if the person had no phone number at the time the schedule was created

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#4512]
Seen on beta.dev, 2.15.0-beta.11, logged in as sharon.

To reproduce:
1) Send a pregnancy registration from the web app (use full access or a restricted user logged in a level above the CHW area). Make sure the contact associated with your user has a phone number.
2) Go to reports and check the schedule is generated. Notice that the outgoing messages are going to the contact associated with your user (known config issue that will be resolved).
3) Change the phone number of the contact associated with your user.
4) Go back to reports and notice that the phone number for outgoing messages in your schedule hasn't updated.
5) Delete the phone number for the contact associated with your user and repeat 1-4 (this time, instead of changing the phone number, just add a phone number where there wasn't one). Notice that when you add a phone number, the schedule updates to that number. Change the phone number and notice the schedule again updates correctly.

**Comment by garethbowen** [comment]
@sglangevin I wonder if the correct solution is to change the config. It's not well defined but the code suggests that "reporting_unit" should always go to whoever sent the message. This is most useful for auto replies and the like. However, I think scheduled messages should go to "clinic" which is the report's clinic's contact's number.

For example, if a nurse registers a pregnancy then they should get an auto reply (configured using "reporting_unit") but the CHW should get the scheduled messages (configured using "clinic").

**Comment by garethbowen** [comment]
@abbyad I'm interested in your opinion here too - what should "reporting_unit" resolve to?

**Comment by sglangevin** [comment]
Sorry I missed this comment yesterday - I thought I had gotten back to you on all but I guess not! In this case, I think the phone number should update regardless of who is receiving the messages. Yes, in most cases we do want SMS messages to go to the CHW, but we often also have messages that go to the CHW's supervisor in pregnancy schedules and I would expect those phone numbers to update in real time as well. Are you suggesting that we just use a different way of indicating the recipient other than `reporting_unit`? I'm not as familiar with SMS config so perhaps @abbyad can give a more helpful response.

**Comment by garethbowen** [comment]
That's exactly what I'm suggesting - the way `reporting_unit` has always worked is it goes to the phone number that sent it originally. We could change the implementation to instead send the message to the contact but this wouldn't work for cases where the sender is unknown and therefore no contact is found. I'm concerned about changing the way it works because that'll change the behaviour for existing projects.

There are a few different ways to configure the recipient to send to the contact depending on exactly what workflow you're implementing.

**Comment by abbyad** [comment]
> "reporting_unit" should always go to whoever sent the message.

I think that is accurate. I was planning on updating the Standard config accordingly.

**Comment by garethbowen** [comment]
Ok, so it sounds like we can close this for acceptance testing again to make sure "clinic" or some other configuration works as expected?

**Comment by sglangevin** [comment]
So are we saying that we can no longer use "reporting_unit" in configurations like this one? Because my assumption would be that regardless of which phone number the message is being sent to, it should be updated correctly. What about messages that have to go to the supervisor? Does this feature only work when we are updating the CHW's phone number? I think I'm confused about what we do and do not support here.

**Comment by sglangevin** [comment]
So are we saying that we can no longer use "reporting_unit" in configurations like this one? Because my assumption would be that regardless of which phone number the message is being sent to, it should be updated correctly. What about messages that have to go to the supervisor? Does this feature only work when we are updating the CHW's phone number? I think I'm confused about what we do and do not support here.

**Comment by garethbowen** [comment]
@sglangevin This is how I'd configure it. NB: I haven't tested this so YMMV.

| Recipient | Resolves to | Useful for |
|---|---|---|
| reporting_unit | `doc.sms_message.from` (never updated) | Auto replies |
| clinic | `doc.contact.parent.contact.phone` | Scheduled messages to the CHP |
| health_center | `doc.contact.parent.parent.contact.phone` | Scheduled messages to the supervisor |

Have you got a use case outside of those three?

You're after a new recipient keyword (eg: "reporter") or perhaps just extending the functionality of the  "reporting_unit". When a report comes in we resolve the sender's contact doc and store it against the report. For SMS this would be resolved using the `from` phone number. For Enketo the sender would be the user that created the doc. Either way, this may or may not be the same as the person the report is currently "assigned" to (ie: the `doc.contact`). Then specifying a recipient of "reporter" would resolve to `sender.phone` falling back to `sms_message.from` when no sender is found.

This is just off the top of my head and would need more design but I think this is outside the scope of 2.15.0.

**Comment by sglangevin** [comment]
Yeah, I don't want to add more work, just want to make sure that I understand so I can document things well for release notes. 

What does reporting_unit resolve to when the sender is a web app user? It sounds like our SMS configs may need to be updated to use `doc.contact` everywhere to make use of the just-in-time phone number updates. Is that correct? I'd like to be clear in the release notes what changes need to be made. We'll handle it centrally for Standard but for our Nepal MoH projects we may need to make changes to the config because most of our configs are using the reporting_unit as the recipient of scheduled messages. However, for Nepal MoH, the reporting_unit is always the CHW. Are we saying that even for a config like that, where the CHW is the reporting_unit (person registering a pregnancy) we need to change it to `doc.contact.parent.contact.phone` to make sure the phone number updates in real time after being changed?

**Comment by garethbowen** [comment]
> What does reporting_unit resolve to when the sender is a web app user?

It falls back to the `doc.contact.phone`. If that's not defined then I think it will just say "reporting_unit" in the UI.

> It sounds like our SMS configs may need to be updated to use doc.contact everywhere to make use of the just-in-time phone number updates. Is that correct?

It depends who you want the message to go to. Use the table above to update the config appropriately.

> I'd like to be clear in the release notes what changes need to be made. 

To be fair, the current behaviour is completely consistent with the old behaviour - all existing projects will keep working the way they always have. We have added a new feature to SMS the assigned contact which can be used instead of "reporting_unit" which sends to the phone number the report was received from. So nothing _needs_ to change unless they want to use this new feature. I'm happy to write the release notes if you think that'd help.

> Are we saying that even for a config like that, where the CHW is the reporting_unit (person registering a pregnancy) we need to change it to doc.contact.parent.contact.phone to make sure the phone number updates in real time after being changed?

Just use "clinic" rather than that long horrible fragile string.

**Comment by sglangevin** [comment]
Ok - did some more testing on beta-old.dev because beta.dev is still down. I uploaded a config where groups 1 & 3 of the ANC reminders have a recipient of 'clinic' and groups 2 & 4 have a recipient of 'health_center'. However, all of the reminders are listed as going to the person who submitted the report, except for the note to the CHW in the case of a pregnancy form being submitted via the app.

[Example 1](https://beta-old.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/251cb6a6-2c5f-4edb-947b-577c57b78b14): Submitted a pregnancy registration via the app

[Example 2](https://beta-old.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/dd32e2d08c7def675f5de6ab96046919): Submitted a pregnancy registration via SMS using the number of the CHW whose area the patient belongs to.

Have I misconfigured or is there a bug somewhere?

**Comment by garethbowen** [comment]
Ooookay... I had a look at example 1 (thanks for the links!) and the problem is the report doesn't have an associated CHP. The contact of the report is Nurse Mary, which makes sense seeing she's the one who submitted it. When you say the recipient is "clinic" it goes up the contact tree from Nurse Mary to     ABC Health Center to District Hospital and if none of those have type "clinic" (spoiler alert: they don't) then it falls back to `doc.from` again. I don't think there's any way to get the patient's CHP until #4442 is fixed, at which point you could use `patient.parent.contact.phone` (off the top of my head). Now I'm convinced this is a straight up bug, because when you say "clinic" you always mean the patient's clinic not the sender's clinic (whatever that means). We should probably change the implementation to try the patient's clinic and fall back to the sender's clinic for backwards compatibility.

TL;DR: I think we should change the resolution to be patient centric not sender centric.

@sglangevin Does that make sense to you?

**Comment by garethbowen** [comment]
Code review please @SCdF 

I merged the `2.15.x` version so I could get a beta out for AT. If you could please review the above PR for the forward port.

**Comment by sglangevin** [comment]
@garethbowen I tried listing clinic and health_center as recipients for different message groups. The clinic recipient resolved to the correct phone number but the health_center recipient resolved to the health_center object. I'm assuming that health_center should resolve to a phone number as well, correct? Check it out [here](https://beta-old.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/dd32e2d08c7def675f5de6ab960a06da).

**Comment by garethbowen** [comment]
The recipient aliases were actually "parent" and "grandparent" but they're not very good, so I've added "health_center" and "district" to be released in 2.15.0-beta.14. Once you upgrade the existing messages should show correctly.

**Comment by sglangevin** [comment]
LGTM. 'clinic' & 'health_center' are working well and are now patient-centric. I'll document that recipient resolution is now patient-centric and list the recipient aliases that we now have. The messages showed correctly after upgrade.

@abbyad we'll probably want to update the recipients in standard config to use this new format so that we make sure that phone number are updating correctly. Using reporting_unit will still be supported but the phone numbers won't necessarily update in real time.

**Comment by sglangevin** [comment]
Ok - I'm still seeing some weird behavior. Once I tried to update the phone number for the CHW ('clinic'), the phone number was replaced by a phone number for someone in a different part of the hierarchy. [See this record](https://beta-old.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/347b3251-5f32-4ea8-86ee-844032fd01d7). Groups 2 & 4 are correct, but groups 1 & 3 are going to a seemingly random number when they should be going to the CHW's phone number. Also, the auto-reply is going to the CHW's old number, not sure if that is related.

**Comment by garethbowen** [comment]
The problem stems from the fact that this code is so old the `parent` is aliased to the `health_center` from back before people existed.

The fix is to not use the extendedTemplateContext for phone number resolution unless all other methods have failed.

The auto-reply is probably not updating because it's in the `pending` state at which point the message content and phone number are fixed ready to be picked up by gateway.

**Comment by sglangevin** [comment]
I wasn't expecting the auto-reply to update - if you look at the CHW Phone field in the report, you'll notice that it is different from the phone number that the auto-reply was sent to.

**Comment by garethbowen** [comment]
Who is the configured recipient for that autoreply? I think the configuration has changed because I can't find that message in use anywhere...

**Comment by sglangevin** [comment]
Hmmm, that's weird. It should be set to the CHW's phone number that is pulled into the form - `chw_phone`. I don't think that has changed, but I did try re-uploading the config at some point during testing, so maybe someone else had changed the config in the interim? This part of the config hadn't changed though, it is what we've had in the Standard config all along.

**Comment by garethbowen** [comment]
Ok, bug number 2: the template context is built up from patient, then previous registrations, then the doc. Because there is a previous registration it's picking up the chw_phone from the previous one instead of this one. I think if I switch the order to patient, doc, _then_ previous registrations it'll just work.
